### PR TITLE
Web Extensions Declarative Net Request WIP

### PR DESCRIPTION
**Pull request by jdatapple**
#### Summary
A work in progress branch for adding Declarative Net Request API and manifest support for Web Extensions.

### CODE REVIEW

**Comment by Elchi3** [comment]
Looks alright to me but would appreciate if @rebloor, @Rob--W or @rpl would take a look since I don't know WebExtensions all that well :)

**Comment by rebloor** [comment]
It's a little difficult to comment without some context on this. Also, have you run `npm test` and confirmed that it has reported no errors.

**Comment by rpl** [comment]
@Rob--W would you mind to provide some feedback on this? (in its current form it is mainly a skeleton of the structure meant to track DeclarativeNetRequest implementation across browser vendors, while the `version` values associated to each browser are not yet populated).

File: webextensions/api/declarativeNetRequest.json [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+              },
+              "safari_ios": {
+                "version_added": false
+              }
+            }
+          }
+        },
+        "onRuleMatchedDebug_event": {
+          "__compat": {
+            "mdn_url": "https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/onRuleMatchedDebug_event",
```
[the remark this answers is not in the document]
Comment (queengooborg) [reply webextensions/api/declarativeNetRequest.json:175]: Actually, this should be the other way around!  The `on` prefix should be removed but the `_event` suffix should stay.

[the remark this answers is not in the document]
Comment (jdatapple) [reply webextensions/api/declarativeNetRequest.json]: Addressed in 1ca88f3.

File: webextensions/declarative_net_request/rules.json [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+                  "opera": {
+                    "version_added": false
+                  },
+                  "safari": {
+                    "version_added": false
+                  }
+                }
+              }
+            },
+            "allowallrequests": {
```
[the remark this answers is not in the document]
Comment (jdatapple) [reply webextensions/declarative_net_request/rules.json]: Addressed in 1ca88f3.

File: webextensions/declarative_net_request/rules.json [hunk 3]
```diff
@@ -0,0 +1,590 @@
+{
+  "webextensions": {
+    "declarative_net_request": {
```
[the remark this answers is not in the document]
Comment (jdatapple) [reply webextensions/declarative_net_request/rules.json]: Addressed in 1ca88f3.

File: webextensions/declarative_net_request/rules.json [hunk 4]
```diff
@@ -0,0 +1,590 @@
...
+      "rules": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/declarative_net_request/rules",
```
[the remark this answers is not in the document]
Comment (jdatapple) [reply webextensions/declarative_net_request/rules.json]: Addressed in 1ca88f3.

**Comment by jdatapple** [comment]
> Here's some feedback on the structure. It probably makes sense to do a full review once the version compatibility info is added + MDN docs are ready for review too.

I believe [1ca88f3](https://github.com/mdn/browser-compat-data/pull/14942/commits/1ca88f3539a0aa128de93fda81ddedf733d80899) addresses your feedback. Is this ready enough for me to start collecting browser data?

**Comment by jdatapple** [comment]
@Rob--W or @queengooborg any other feedback on this so that we can continue to make progress?

**Review by Rob--W — CHANGES_REQUESTED** [review]
> Is this ready enough for me to start collecting browser data?

By this, are you referring to `version_added`? With the changes below, the answer is yes.

You can skip Firefox for now, DNR is still under development there and there is no version that supports it yet.

**Comment by queengooborg** [comment]
I happened to notice that every single feature here has no implementation yet.  We've actually implemented a policy _not_ to include features with no support in any browser in BCD to reduce clutter.  Are these features going to be implemented in Safari or other browsers?  Also, is this PR still planned?

**Comment by Rob--W** [comment]
> I happened to notice that every single feature here has no implementation yet.  We've actually implemented a policy _not_ to include features with no support in any browser in BCD to reduce clutter.  Are these features going to be implemented in Safari or other browsers?  Also, is this PR still planned?

What lead to the conclusion that these features have no implementation?

There is an implementation in Chrome and Safari. And at the rate that the PR progresses, there may even be an implementation in Firefox when the PR+docs completes.

**Comment by queengooborg** [comment]
> What lead to the conclusion that these features have no implementation?

In this PR, every feature added is set to `false` for all browsers.  I realize that there's been comments that clarify there's support; my apologies!

Since this PR has been sitting around for a long time, though, would you be down to create a new PR to supersede this one, @Rob--W?

**Comment by Rob--W** [comment]
> Since this PR has been sitting around for a long time, though, would you be down to create a new PR to supersede this one, @Rob--W?

Despite the lack of visible progress on the PR, the PR has not been forgotten. Once the supported features have been clarified, and documentation for MDN is ready, I would expect more progress here.

If somehow the PR hasn't finished when Firefox's implementation ships, then I may consider taking over. But for now I think the PR is in good hands.

**Comment by Rob--W** [comment]
This PR can be closed, because it has been superseded by #18403.

**Pull request closed without merging**